core/server/controllers/frontend/index.js

```diff
@@ -66,11 +66,21 @@ frontendControllers = {
                 return next();
             }
 
-            // If we're ready to render the page but the last param is 'edit' then we'll send you to the edit page.
+            // CASE: we only support /:slug format for pages
+            if (post.page && post.url !== req.path) {
+                return next();
+            }
+
+            // CASE: last param is of url is /edit, redirect to admin
             if (lookup.isEditURL) {
                 return res.redirect(config.paths.subdir + '/ghost/editor/' + post.id + '/');
             }
 
+            // CASE: permalink is not valid anymore, we redirect him to the correct one
+            if (post.url !== req.path) {
+                return res.redirect(post.url);
```

Comment:
Yep, that is what I'm saying 😉  this should be a 301.

The default if you just use `res.redirect` is a 302: http://expressjs.com/en/4x/api.html#res.redirect